text/0000-command-ergonomics.md

```diff
@@ -0,0 +1,634 @@
+
+ 3. Not include a `cause` at all.
+    This is the option we propose.
+
+## `read_stdout` etc. naming.
+
+We have `fs::read_to_string`.
+
+Possible names (taking `read_stdout_string` as the example):
+
+ * `read_stdout_string` (proposed in this RFC)
+ * `stdout_string` but `stdout` is alread taken.
+ * `get_stdout_string`
+ * `run_get_stdout_string`
+ * `run_stdout_string`
+
+It is difficult to convey everything that is needed in a short name.
+In particular, all of these functions spawn the program,
+and wait (at an appropriate point) for it to exit.
+
+Should `read_output` be the one that returns `Vec<u8>`
+or the one that returns `Vec<String>` ?
+Precedent in the stdlib is usually to have the bytes version undecorated
+(eg, `fs::read_to_string`).
+
+## stderr handling default
+
+What should happen if the Rust programmer didn't call `.stderr()` ?
+
+Options are:
+
+ 1. Treat it as `inherit`.
+   This is probably fine for a command line application,
+   but may be too lax for many other contexts.
+
+ 2. Treat it as `piped`: call any stderr output an error.
+   This is a fairly conservative choice,
+   but it can lead to unexpected runtime errors,
+   if a called program later starts printing warning messages.
+
+ 3. Call this a programming mistake, and panic.
+
+ 4. Somehow make this a compile error.
+   This would involve a new `CommandForWhichWeHaveSpecifiedStderr`
+   type (typestate pattern), or providing the stderr handling as a mandatory argument
+   to all the new methods.
+   These seem unpalatably unergonomic.
+
+5. One of the above, and additionally provide a lint that makes a best-effort attempt to catch this at compile time.
+
+Here we propose option 1: treat as `inherit`.
+
+# Examples
+
+## Conveniently running `diff`
+
+```rust
+    let result = Command::new("diff")
+        .args(["before","after"])
+        .stderr(Stdio::piped()) // optional, could just let it inherit
+        .run();
+    let status = match result {
+        Ok(()) => 0,
+        Err(err) => err.just_status()?.code(),
+    };
+```
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+## Providing a way to combine and interleave stdout and stderr
+
+Currently, `Command` insists on separating out stdout and stderr,
+if you ask to capture them.
+If you want them combined
+(which is the only way to preserve the relative ordering)
+you must do one of:
+
+ * run a command which itself does the redirection
+   (easy using the shell on Unix)
+
+ * send them each to your own stdout/stderr with `inherit`
+   and expect your caller to combine them
+
+ * sent themk to the *same* one of your stdout/stderr
+   which will be possible after
+   https://github.com/rust-lang/rust/pull/88561)
+
+We should provide something like this:
+
+```
+impl Command {
+    /// Arranges that the command's stderr will be sent to wherever its stdout is going
+    fn stderr_to_stdout();
+}
+```
+
+(It is not sensibly possible at least on Unix
```

Comment:
```suggestion
(It is not easy on Unix targets
```

It is possible, though quite awkward and with some limitations. (Even more limitations if you're not on Linux.) I needed a solution to this, and ended up writing [io-mux](https://crates.io/crates/io-mux) for this.